File: backend/src/main/kotlin/com/alpha/web/controller/CategoryController.kt
package com.alpha.web.controller

import com.alpha.service.CategoryService
import com.alpha.service.dto.*
import com.alpha.web.dto.ApiResponse
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.tags.Tag
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.util.*

@RestController
@RequestMapping("/api/v1/categories")
@Tag(name = "Categories", description = "Category management endpoints")
class CategoryController(
    private val categoryService: CategoryService
) {
    
    @GetMapping
    @Operation(summary = "Get all active categories")
    fun getAllCategories(): ResponseEntity<ApiResponse<List<CategoryResponse>>> {
        val categories = categoryService.getAllCategories()
        return ResponseEntity.ok(ApiResponse.success(categories))
    }
    
    @GetMapping("/{categoryId}")
    @Operation(summary = "Get category by ID")
    fun getCategory(
        @PathVariable categoryId: UUID
    ): ResponseEntity<ApiResponse<CategoryResponse>> {
        val category = categoryService.getCategory(categoryId)
        return ResponseEntity.ok(ApiResponse.success(category))
    }
    
    @PostMapping
    @Operation(summary = "Create new category (admin only)")
    fun createCategory(
        @Valid @RequestBody request: CreateCategoryRequest
    ): ResponseEntity<ApiResponse<CategoryResponse>> {
        val category = categoryService.createCategory(request)
        return ResponseEntity.status(HttpStatus.CREATED)
            .body(ApiResponse.success(category))
    }
    
    @PutMapping("/{categoryId}")
    @Operation(summary = "Update category (admin only)")
    fun updateCategory(
        @PathVariable categoryId: UUID,
        @Valid @RequestBody request: UpdateCategoryRequest
    ): ResponseEntity<ApiResponse<CategoryResponse>> {
        val category = categoryService.updateCategory(categoryId, request)
        return ResponseEntity.ok(ApiResponse.success(category))
    }
    
    @DeleteMapping("/{categoryId}")
    @Operation(summary = "Delete category (admin only)")
    fun deleteCategory(
        @PathVariable categoryId: UUID
    ): ResponseEntity<ApiResponse<Unit>> {
        categoryService.deleteCategory(categoryId)
        return ResponseEntity.ok(ApiResponse.success())
    }
}

File: backend-scala-zio/src/main/scala/com/alpha/controller/CategoryController.scala
package com.alpha.controller

import zio.*
import zio.http.*
import zio.json.*
import com.alpha.service.*
import com.alpha.dto.*
import com.alpha.domain.model.*
import java.util.UUID

class CategoryController(categoryService: CategoryService):

  private def jsonResponse[T: JsonEncoder](data: ApiResponse[T]): Response =
    Response.json(JsonEncoder[ApiResponse[T]].encodeJson(data, None))

  private def errorResponse(message: String, status: Status): Response =
    Response.text(message).status(status)

  val routes = Routes(
    Method.GET / "api" / "v1" / "categories" -> handler(handleGetAll()),
    Method.GET / "api" / "v1" / "categories" / string("id") -> handler(handleGetById),
    Method.POST / "api" / "v1" / "categories" -> handler(handleCreate),
    Method.PUT / "api" / "v1" / "categories" / string("id") -> handler(handleUpdate),
    Method.DELETE / "api" / "v1" / "categories" / string("id") -> handler(handleDelete)
  )

  private def handleGetAll(): Task[Response] =
    categoryService.getAllCategories
      .map(categories => jsonResponse(ApiResponse.successList(categories)))
      .catchAll(error => ZIO.succeed(errorResponse(error.getMessage, Status.InternalServerError)))

  private def handleGetById(id: String, req: Request): Task[Response] =
    ZIO.attempt(UUID.fromString(id)).flatMap { uuid =>
      categoryService.getCategory(uuid).map {
        case Some(category) => jsonResponse(ApiResponse.success(category))
        case None => errorResponse("Category not found", Status.NotFound)
      }
    }.catchAll(error => ZIO.succeed(errorResponse(error.getMessage, Status.BadRequest)))

  private def handleCreate(request: Request): Task[Response] =
    (for
      body <- request.body.asString
      req <- ZIO.fromEither(JsonDecoder[CreateCategoryRequest].decodeJson(body)).orElseFail(new Exception("Invalid request body"))
      created <- categoryService.createCategory(req)
    yield jsonResponse(ApiResponse.success(created)))
      .catchAll(error => ZIO.succeed(errorResponse(error.getMessage, Status.BadRequest)))

  private def handleUpdate(id: String, request: Request): Task[Response] =
    (for
      body <- request.body.asString
      req <- ZIO.fromEither(JsonDecoder[UpdateCategoryRequest].decodeJson(body)).orElseFail(new Exception("Invalid request body"))
      uuid <- ZIO.attempt(UUID.fromString(id))
      updated <- categoryService.updateCategory(uuid, req)
    yield jsonResponse(ApiResponse.success(updated)))
      .catchAll(error => ZIO.succeed(errorResponse(error.getMessage, Status.BadRequest)))

  private def handleDelete(id: String, req: Request): Task[Response] =
    ZIO.attempt(UUID.fromString(id)).flatMap { uuid =>
      categoryService.deleteCategory(uuid).as(Response.status(Status.NoContent))
    }.catchAll(error => ZIO.succeed(errorResponse(error.getMessage, Status.BadRequest)))

object CategoryController:
  val layer: ZLayer[CategoryService, Nothing, CategoryController] =
    ZLayer.fromFunction(new CategoryController(_))
